light/trie.go

```diff
@@ -0,0 +1,126 @@
+// Copyright 2015 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package light
+
+import (
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/ethdb"
+	"github.com/ethereum/go-ethereum/trie"
+	"golang.org/x/net/context"
+)
+
+type LightTrie struct {
+	trie         *trie.SecureTrie
+	originalRoot common.Hash
+	odr          OdrBackend
+	db           ethdb.Database
+}
+
+func NewLightTrie(root common.Hash, odr OdrBackend, useFakeMap bool) *LightTrie {
+	return &LightTrie{
+		// SecureTrie is initialized before first request
+		originalRoot: root,
+		odr:          odr,
+		db:           odr.Database(),
+	}
+}
+
+// retrieveKey retrieves a single key, returns true and stores nodes in local
+// database if successful
+func (t *LightTrie) retrieveKey(ctx context.Context, key []byte) bool {
+	r := &TrieRequest{ctx: ctx, root: t.originalRoot, key: key}
+	return t.odr.Retrieve(r) == nil
+}
+
+// compactHexEncode encodes a series of nibbles into a byte array
+func compactHexEncode(nibbles []byte) []byte {
+	nl := len(nibbles)
+	if nibbles[nl-1] == 16 {
+		nl--
+	}
+	l := (nl + 1) / 2
+	var str = make([]byte, l)
+	for i, _ := range str {
+		b := nibbles[i*2] * 16
+		if nl > i*2 {
+			b += nibbles[i*2+1]
+		}
+		str[i] = b
+	}
+	return str
+}
+
+func (t *LightTrie) do(ctx context.Context, fallbackKey []byte, fn func() error) error {
+	err := fn()
+	for err != nil {
+		mn, ok := err.(*trie.MissingNodeError)
+		if !ok {
+			return err
+		}
+		nibbles := append(mn.KeyPrefix, mn.KeySuffix...)
+		var key []byte
+		if len(nibbles) > 0 {
+			key = compactHexEncode(nibbles)
```

Comment:
This is weird. If we need the full encoded key, we might just as well put it in the error when it is created.